lib/ndebug/user.c

```diff
@@ -0,0 +1,176 @@
+    }
+
+    while (true) {
+        ssize_t bytes = ndebug_usb_read(NDEBUG_CHANNEL_USR,
+                                        NDEBUG_MAX_PACKET_SIZE,
+                                        timeout, scratch_buffer);
+        if (bytes < 0) {
+            return bytes;
+        } else if (bytes < (ssize_t)sizeof(ndebug_ctrl_packet_t)) {
+            continue;
+        }
+
+        if (is_valid_connection_request(bytes, scratch_buffer)) {
+            // Connection established.
+            if (send_message_to_host(NDEBUG_CTRL_CMD_ESTABLISHED) < 0) {
+                continue;
+            }
+            connected = true;
+            return NO_ERROR;
+        } else {
+            // Tell the host that it needs to reset the connection before
+            // attempting any further communication.
+            send_message_to_host(NDEBUG_CTRL_CMD_RESET);
+        }
+    }
+}
+
+ssize_t ndebug_read_usr(uint8_t *buf, const lk_time_t timeout)
+{
+    LTRACEF("buf: 0x%p, timeout: %u\n", buf, timeout);
+
+    if (!connected) {
+        // User must call ndebugusr_await_host and establish a connection
+        // before continuing.
+        return ERR_CHANNEL_CLOSED;
+    }
+
+    mutex_acquire(&usr_mutex);
+
+    // Retry Loop
+    while (true) {
+        ssize_t result = ndebug_usb_read(NDEBUG_CHANNEL_USR,
+                                         NDEBUG_MAX_PACKET_SIZE,
+                                         timeout, scratch_buffer);
+        if (result < (ssize_t)sizeof(ndebug_ctrl_packet_t)) {
+            dprintf(INFO, "Short Packet. Len = %ld\n", result);
+            continue;
+        }
+
+        ndebug_ctrl_packet_t *ctrl = (ndebug_ctrl_packet_t *)scratch_buffer;
+        if (ctrl->magic != NDEBUG_CTRL_PACKET_MAGIC) {
+            dprintf(INFO, "Bad Packet Magic = %u\n", ctrl->magic);
+            continue;
+        }
+
+        if (ctrl->type == NDEBUG_CTRL_CMD_RESET) {
+            send_message_to_host(NDEBUG_CTRL_CMD_RESET);
+            connected = false;
+            mutex_release(&usr_mutex);
+            return ERR_CHANNEL_CLOSED;
+        } else if (ctrl->type == NDEBUG_CTRL_CMD_DATA) {
+            const ssize_t n_data_bytes = result - sizeof(ndebug_ctrl_packet_t);
+            memcpy(buf, scratch_buffer + sizeof(ndebug_ctrl_packet_t),
+                   n_data_bytes);
+            mutex_release(&usr_mutex);
+            return n_data_bytes;
+        } else {
+            dprintf(INFO, "Unexpected packet type = %u\n", ctrl->type);
+        }
+    }
+}
+
+ssize_t ndebug_write_usr(uint8_t *buf, const size_t n, const lk_time_t timeout)
+{
+    LTRACEF("buf = 0x%p, n = %u, timeout = %u\n", buf, n, timeout);
+
+    if (!connected) {
+        // User must call ndebugusr_await_host and establish a connection
+        // before continuing.
+        return ERR_CHANNEL_CLOSED;
+    }
+
+    // TODO(gkalsi): Packetize the packet that's too big and send the data
+    //               anyway.
+    if (n > NDEBUG_USR_MAX_PACKET_SIZE)
```

Comment:
I think this is fine because `NDEBUG_USR_MAX_PACKET_SIZE` is defined as `MAX_PACKET_SIZE - sizeof (header)`